### PR TITLE
Add method for creating immutable DTOs

### LINKED ISSUE #4: Support creating immutable DTOs
We could add `CreateNew(IEnumerable<(string, object?)> properties)` to `IDtoInfo`. It would create an instance of the DTO, assigning the specified values to properties of the specified names.

This would be useful for mutable DTOs, but critical for immutable DTOs, which it could support by calling a constructor whose parameter names match read-only property names.

The introduction of immutable records with positional parameters in C# 9 makes this more pressing.

We should respect any default parameter values on the constructor.

We should support mixed mutable/immutable properties.

All property and parameter name matching should be case-insensitive.

If there are multiple possible paths, we should probably prefer constructor parameters to settable properties, for the sake of efficiency.

Adding `IDtoInfo.Constructors` could be useful, both for this behavior and for clients. Maybe detect static "constructor methods" as well?

### PR DESCRIPTION
**Pull request by TyMick**
How's this look to close #4? To summarize, this `CreateNew` method first uses [`InvokeMember`](https://docs.microsoft.com/en-us/dotnet/api/system.type.invokemember?view=netcore-3.1#System_Type_InvokeMember_System_String_System_Reflection_BindingFlags_System_Reflection_Binder_System_Object_System_Object___System_Reflection_ParameterModifier___System_Globalization_CultureInfo_System_String___) to try and find a constructor that takes _all_ of the passed items. If it doesn't find one, it'll separate out all the items with names and types that match those of mutable properties/fields of the DTO type, then look for a constructor that takes the remaining items, and if it's successful, it'll set the mutable props/fields manually after creating the instance.

I named the parameter `argsAndProps` instead of `properties` when I realized that constructor arguments could have different names from the immutable properties they set, but I can switch back to `properties` or something else if it would look nicer. I just can't think of a way to target the immutable property names directly since the abstraction of a constructor method conceals which arguments affect which immutable properties.

I was pleasantly surprised to find that the [`DefaultBinder`](https://docs.microsoft.com/en-us/dotnet/api/system.type.defaultbinder?view=netcore-3.1) already matches named parameters in a case-insensitive manner, so I didn't have to write a new `Binder`. 👍🏼

Would adding an `IDtoInfo.Constructors` still be useful? I wrote most of a `DtoConstructor` class (inspired by `DtoProperty`) before I realized that I didn't need it if I used `InvokeMember`, so I could finish that up if it'd be handy.

### CODE REVIEW

**Comment by ejball** [comment]
Thanks! I will look at this when I can. I've also been playing around with issue #4 but haven't pushed anything up yet.

File: src/Faithlife.Reflection/DtoInfo.cs [hunk 1]
```diff
@@ -108,6 +108,65 @@ public sealed class DtoInfo<T> : IDtoInfo
...
+		{
+			try
+			{
+				// Try for a constructor that accepts all passed arguments
+				string[] argNames = argsAndProps.Select(arg => arg.Name).ToArray();
+				object?[] argValues = argsAndProps.Select(arg => arg.Value).ToArray();
+				var flags = BindingFlags.Public | BindingFlags.Instance | BindingFlags.CreateInstance;
+				return (T) typeof(T).InvokeMember(typeof(T).Name, flags, null, null, argValues, null, null, argNames);
+			}
+			catch (MissingMethodException)
```
Comment (ejball) [inline src/Faithlife.Reflection/DtoInfo.cs:127]: For what it's worth, throwing and catching exceptions as part of normal code flow is bad for performance and generally not a good practice.

Comment (TyMick) [reply src/Faithlife.Reflection/DtoInfo.cs:127]: Ah, the Python community got me into a bad habit in that way, then, haha.

Now that I'm looking for alternatives, though, I'm not finding any other reflection methods that can make intelligent matches using parameter names like [`InvokeMember`](https://docs.microsoft.com/en-us/dotnet/api/system.type.invokemember?view=netcore-3.1) can. [`GetMember`](https://docs.microsoft.com/en-us/dotnet/api/system.type.getmember?view=netcore-3.1), [`GetMethod`](https://docs.microsoft.com/en-us/dotnet/api/system.type.getmethod?view=netcore-3.1), and [`GetConstructor`](https://docs.microsoft.com/en-us/dotnet/api/system.type.getconstructor?view=netcore-3.1) all require knowledge of what order to put the parameters in, and when I [search the .NET docs for "namedParameters"](https://docs.microsoft.com/en-us/search/?terms=namedParameters&scope=.NET), all that comes up is various incarnations of `InvokeMember`.

Should I go ahead and finish that `DtoConstructor` class and add functionality to test a constructor to find if it's compatible given parameter names and types? That way we could find out whether `InvokeMethod` will work before calling it, so we wouldn't have to catch an exception. Or would that be too much reinventing of the wheel?

Comment (ejball) [reply src/Faithlife.Reflection/DtoInfo.cs:127]: I didn't even know `InvokeMember` could do that!

I've pushed my work to this PR: https://github.com/Faithlife/FaithlifeReflection/pull/7

It maintains a dictionary of parameter indices to get the order right.

It doesn't expose the constructors in the API, but it does maintain a list of compatible constructors ("creators") internally.

Comment (TyMick) [reply src/Faithlife.Reflection/DtoInfo.cs:127]: Ah yeah, that'll do the trick! So no need to worry about constructors with parameter names that don't match the properties they affect?

Come to think of it, I suppose I don't need to assume we don't know _anything_ about the DTOs this class will encounter—given their purpose of "simply" transferring data, it's probably pretty safe to assume there's no reason for their constructors to do anything but directly assign values to properties, in which case there's no reason to expect parameter names to be any different from property names (except in letter casing).

Comment (TyMick) [reply src/Faithlife.Reflection/DtoInfo.cs:127]: Should I go ahead and close this PR? 👍🏼

Comment (ejball) [reply src/Faithlife.Reflection/DtoInfo.cs:127]: > constructors with parameter names that don't match the properties they affect

That PR rejects any constructor with a parameter name that doesn't match a property. But we do assume that if the names match, it just sets the property.

> Should I go ahead and close this PR?

You bet. Thanks!

**Pull request closed without merging**